plugins/formal-verification/testData/diagnostics/no_contracts/full_viper_dump.fir.diag.txt

```diff
@@ -292,6 +292,13 @@ field class_Foo$member_x: Int
 
 field special$function_object_call_counter: Int
 
+function T_class_global$class_Foo$get$class_Foo$member_x(get$function$subject: Ref): Int
+  requires acc(T_class_global$class_Foo(get$function$subject), write)
```

Comment:
At the moment, all these predicates are `write`.  Do we ever expect to have non-write predicates here?  E.g. if we have read access to `a.x`, do we expect to still have `write` access to the class predicate of `a.x`?
(Actually, now that I see this, I wonder whether the RO vs non-RO predicate stuff could be solved by tracking permission amounts...  We should think about this in more detail.)

Reply:
I've never thought about it because `acc(T_class_global$class_Foo(get$function$subject), write)` is just an equivalent syntax for `T_class_global$class_Foo(get$function$subject)`.
I think that `acc(T_class_global$class_Foo(get$function$subject), wildcard)` gives you only read permissions to all the accesses in the predicate while `acc(T_class_global$class_Foo(get$function$subject), write)` gives you the permission written in the predicate

Reply:
I think the way it works is multiplicative:

```
predicate P(this) { acc(this.x, 1/3) }
acc(P(a), 1/2)
```

Now unfolding the predicate gives `acc(this.x, 1/6)`.

That said, this isn't exactly what we want with mutable variables because when dealing with those, when we don't have permission with them we shouldn't assume they don't change, either.